### PR TITLE
Add max and min operators

### PR DESCRIPTION
**Pull request by adrianomelo**
they are using reduce to select the min and max elements. maybe a similar approach can be used for sum and average.

### CODE REVIEW

File: src/Observable.ts [hunk 1]
```diff
@@ -187,10 +187,12 @@ export default class Observable<T> implements CoreOperators<T>  {
   map: <T, R>(project: (x: T, ix?: number) => R, thisArg?: any) => Observable<R>;
   mapTo: <R>(value: R) => Observable<R>;
   materialize: () => Observable<any>;
+  max: () => Observable<any>;
   merge: (...observables: any[]) => Observable<any>;
   mergeAll: (concurrent?: any) => Observable<any>;
   mergeMap: <R>(project: ((x: T, ix: number) => Observable<any>), projectResult?: (x: T, y: any, ix: number, iy: number) => R, concurrent?: number) => Observable<R>;
   mergeMapTo: <R>(observable: Observable<any>, projectResult?: (x: T, y: any, ix: number, iy: number) => R, concurrent?: number) => Observable<R>;
+  min: () => Observable<any>;
```
Comment (kwonoj) [inline src/Observable.ts]: is it expected signature does not match with actual implementation?

`min<T, R>(predicate?: (x: R, y: T) => R, thisArg?: R): Observable<R>`

File: src/operators/reduce-support.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+import tryCatch from '../util/tryCatch';
+import {errorObject} from '../util/errorObject';
+
+export class ReduceOperator<T, R> implements Operator<T, R> {
+
+  acc: R;
+  project: (acc: R, x: T) => R;
+
+  constructor(project: (acc: R, x: T) => R, acc?: R) {
```
Comment (kwonoj) [inline src/operators/reduce-support.ts]: field can be assigned directly via ctor like

`ctor(private value1 : any)`

File: src/Observable.ts (hunk 1 above)
Comment (adrianomelo) [inline src/Observable.ts]: Yes, it is.. And I realized that this signature doesn't make sense. I'm using `thisArg` as a `seed` for reduce. Maybe a better signature is `min<T, R>(predicate?: (x: R, y: T) => R): Observable<R>`

Comment (kwonoj) [inline src/Observable.ts]: As RxJS4 does, `comparer` seems bit more clear context for this operator.

Comment (adrianomelo) [inline src/Observable.ts]: I agree :)

File: src/operators/reduce-support.ts (hunk 2 above)
Comment (adrianomelo) [inline src/operators/reduce-support.ts]: @kwonoj I think that it's better leave this change to a new pull request. I didn't changed the code, just moved to a new file. There are some other minor changes such as a better signature, `seed` instead of `acc`. What do you think?

Comment (kwonoj) [inline src/operators/reduce-support.ts]: In my case I do apply small refactoring like this when I create PR, but if there's chunk of changes can be created separate PR that seems also way. Think @blesh can advise way feasible for maintaining repo history.

Comment (blesh) [inline src/operators/reduce-support.ts]: > I think that it's better leave this change to a new pull request. I didn't changed the code, just moved to a new file. 

This is preferable.

**Comment by blesh** [comment]
Looks good, and I'm glad to have these operators.

I think, though, that since both of these operators can be accomplished trivially with the `reduce` operator, we can have them only exist in the `KitchenSink` output for now, and they don't need to be "core".

(Don't worry, I suspect the prevalence of transpiling and the ES7 Function Bind feature are going to make the whole "core" vs "kitchen sink" thing a solved problem)

**Comment by adrianomelo** [comment]
@blesh No problem.. My goal with this patch was understand how the project is structured. So, I chose an easy to implement operator.

**Comment by blesh** [comment]
@adrianomelo ... just let me know when this PR is updated. You'll also probably need to rebase, it looks like there are merge conflicts.

Thanks again for your hard work! I appreciate it.

**Comment by adrianomelo** [comment]
@blesh this PR is now updated.
Thank you for helping me out reviewing code and for maintaining the project. You guys are doing an amazing work.

**Comment by blesh** [comment]
@adrianomelo it is my distinct pleasure.

Your PR is merged with 79cb6cf ... thanks for your hard work!
